clin/clients/nakadi_sql.py

```diff
@@ -0,0 +1,153 @@
+        return (
+            Partitioning(
+                strategy=Partitioning.Strategy(
+                    payload["repartition_parameters"]["partition_strategy"]
+                ),
+                keys=payload["repartition_parameters"].get("partition_key_fields"),
+                partition_count=int(
+                    payload["repartition_parameters"]["number_of_partitions"]
+                ),
+            )
+            if "repartition_parameters" in payload
+            else None
+        )
+
+    return OutputEventType(
+        category=Category(payload["category"]),
+        owning_application=payload.get(
+            "owning_application", event_type.owning_application
+        ),
+        audience=Audience(payload["audience"])
+        if "audience" in payload
+        else event_type.audience,
+        repartitioning=convert_repartitioning(),
+        cleanup=Cleanup(
+            policy=Cleanup.Policy(payload["cleanup_policy"]),
+            retention_time_days=payload.get("retention_time", 0) // MS_IN_DAY,
+        ),
+    )
+
+
+def sql_query_from_payload(event_type: EventType, payload: dict) -> SqlQuery:
+    if payload["id"] != payload["output_event_type"]["name"]:
+        raise NakadiError(
+            "The output event type's name does not match the sql query id."
+            " This is unexpected and unsupported - please report this issue in the clin project on GitHub!"
+        )
+
+    return SqlQuery(
+        name=payload["id"],
+        sql=payload["sql"],
+        envelope=payload["envelope"],
+        output_event_type=output_event_type_from_payload(
+            event_type, payload["output_event_type"]
+        ),
+        auth=auth_from_payload(payload["authorization"]),
+    )
+
+
+def sql_query_to_payload(sql_query: SqlQuery) -> dict:
+    payload = {
+        "id": sql_query.name,
+        "sql": sql_query.sql,
+        "envelope": sql_query.envelope,
+        "output_event_type": {
+            "name": sql_query.name,
+            "owning_application": sql_query.output_event_type.owning_application,
+            "category": str(sql_query.output_event_type.category),
+            "audience": str(sql_query.output_event_type.audience),
+            "cleanup_policy": str(sql_query.output_event_type.cleanup.policy),
+            "retention_time": sql_query.output_event_type.cleanup.retention_time_days
+            * MS_IN_DAY,
+        },
+        "authorization": auth_to_payload(sql_query.auth),
+    }
+
+    if sql_query.output_event_type.repartitioning:
+        payload["repartition_parameters"] = {
+            "number_of_partitions": sql_query.output_event_type.repartitioning.partition_count,
+            "partition_strategy": str(
+                sql_query.output_event_type.repartitioning.strategy
+            ),
+            "partition_key_fields": sql_query.output_event_type.repartitioning.keys,
+        }
+
+    return payload
+
+
+def auth_to_payload(auth: ReadOnlyAuth) -> dict:
```

Comment:
Yes, annoyingly the auth types are slightly different. You're right there's a shared subset and initially I had the shared functionality extracted. I then inlined it because I figured that the shared stuff is actually so trivial that extracting makes the whole function more complex than it's worth. Have you tried extracting and looking at the result? I don't have a super strong opinion and am happy to extract if you really think it's better

Reply:
I will relay on your judgment (=

Reply:
Now that any token is actually supported in both I've refactored again. Much happier with how it looks now!